### PR TITLE
Fixed #24139 -- Evaluate reason_phrase on demand based on status_code.

### PR DESCRIPTION
**Pull request by jdufresne**

### CODE REVIEW

**Comment by shaib** [comment]
buildbot, test this please.

File: django/http/response.py [hunk 1]
```diff
@@ -141,6 +137,16 @@ def charset(self):
     def charset(self, value):
         self._charset = value
 
+    @property
+    def reason_phrase(self):
+        if self._reason_phrase is not None:
```
Comment (MarkusH) [inline django/http/response.py]: You don't set `self._reason_phrase` anywhere in here if the lookup fails. I don't think that's intended?

Comment (collinanderson) [inline django/http/response.py]: It makes sense to me that you would leave it unset unless `self.reason_phase` explicitly sets it.

Comment (MarkusH) [inline django/http/response.py]: Can we have that as a comment then, please.

**Comment by jdufresne** [comment]
> Can we have that as a comment then, please.

Added comment as suggested. Let me know if you had something else in mind.

**Comment by jdufresne** [comment]
Rebased after the merge of #3902 and resolved conflicts.

File: docs/ref/request-response.txt [hunk 2]
```diff
@@ -625,6 +625,12 @@ Attributes
 
     The `HTTP status code`_ for the response.
 
+    .. versionchanged:: 1.9
+
+        Unless :attr:`reason_phrase` is explicitly set, modifying the value of
+        ``status_code`` outside the constructor will also modify the value of
+        `reason_phrase`.
```
Comment (berkerpeksag) [inline docs/ref/request-response.txt]: :attr:\`reason_phrase\` or \`\`reason_phrase\`\`

File: docs/ref/request-response.txt [hunk 3]
```diff
@@ -636,6 +642,11 @@ Attributes
 
         .. _`HTTP standard's`: https://www.ietf.org/rfc/rfc2616.txt
 
+    .. versionchanged:: 1.9
+
+        Unless explicitly set, ``reason_phrase`` is determined by the current
```
Comment (berkerpeksag) [inline docs/ref/request-response.txt]: You can also combine these two versionchanged directives (see line 638).

File: django/http/response.py [hunk 4]
```diff
@@ -37,7 +37,6 @@ class HttpResponseBase(six.Iterator):
     """
 
     status_code = 200
```
Comment (berkerpeksag) [inline django/http/response.py]: Perhaps we can use the same approach for status_code and status attributes, too.

Setting two different attributes to do the same job would be bad:

```py
resp = HttpResponse(status=503)

resp = HttpResponse()
resp.status_code = 503
```

status_code is already used in many places, perhaps we can also add status_code to the constructor of HttpResponseBase.

**Comment by jdufresne** [comment]
> Perhaps we can use a similar approach for status_code and status attributes, too.

Do you mind expanding on what you're requesting?

Are you saying `status_code` should not be a class attribute on `HttpResponseBase`? The other classes in this file rely on this behavior: `HttpResponseRedirect`, `HttpResponseNotFound`, etc.

Or are you saying the `__init__()` kwarg `status` should be renamed to `status_code`?

Sorry, but I'm not sure precisely what "status attributes" is referring to.

I'm happy to cleanup whatever needs it, I'm just not sure exactly what you're asking me to change.

I suppose one potential fix would be to fix the `Content-Type` header to evaluate based on the current value of `self._charset`. Is this along the lines of what you mean?

Version in master (if charset changes outside the constructor, the `Content-Type` header is wrong):

```
    def __init__(self, content_type=None, status=None, reason=None, charset=None):
        ...
        self._charset = charset
        if content_type is None:
            content_type = '%s; charset=%s' % (settings.DEFAULT_CONTENT_TYPE,
                                               self.charset)
```

**Comment by berkerpeksag** [comment]
Sorry for being unclear.

> Or are you saying the __init__() kwarg status should be renamed to status_code?

That's what I meant to say. However, adding a new `status_code` keyword argument would be better for backward compatibility.

**Comment by jdufresne** [comment]
> That's what I meant to say. However, adding a new status_code keyword argument would be better for backward compatibility.

Thanks. Will do.

**Comment by jdufresne** [comment]
I believe I have implemented what you had in mind. Let me know how you feel about the results.

File: tests/template_tests/test_response.py [hunk 5]
```diff
@@ -250,12 +256,19 @@ def test_context_processor_priority(self):
...
     def test_kwargs(self):
         response = self._response(content_type='application/json',
-                                  status=504)
+                                  status_code=504)
         self.assertEqual(response['content-type'], 'application/json')
         self.assertEqual(response.status_code, 504)
 
     def test_args(self):
         response = TemplateResponse(self.factory.get('/'), '', {},
+                                    'application/json', None, None, None, 504)
```
Comment (MarkusH) [inline tests/template_tests/test_response.py]: This raises an error:

```python
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-trusty/database/mysql/label/trusty-pr/python/python2.7/tests/template_tests/test_response.py", line 265, in test_args
    'application/json', None, None, None, 504)
  File "/home/jenkins/workspace/pull-requests-trusty/database/mysql/label/trusty-pr/python/python2.7/django/template/response.py", line 201, in __init__
    RemovedInDjango20Warning, stacklevel=2)
RemovedInDjango20Warning: The current_app argument of TemplateResponse is deprecated. Set the current_app attribute of its request instead.
```

**Comment by timgraham** [comment]
This change seems to have gotten a bit unwieldy as far as the scope of the original ticket goes. Is there really a need to deprecate `status` in favor of `status_code`? I'd try to keep solving this ticket more minimal unless all these changes are really necessary to do together.

**Comment by jdufresne** [comment]
> This change seems to have gotten a bit unwieldy as far as the scope of the original ticket goes. Is there really a need to deprecate status in favor of status_code? I'd try to keep solving this ticket more minimal unless all these changes are really necessary to do together.

I think I agree. In my opinion, there is no pressing need to deprecate `status` in favor of `status_code` (although I agree it is a nice to have for consistency). Shall I discard the 2nd commit? If still preferred I can move the 2nd commit to a different PR.

I originally went down this path at the request of @berkerpeksag. As a core contributor and Django Fellow I interpret his review as authoritative.

If there is disagreement about the 2nd commit, please let me know how to proceed.

**Comment by jdufresne** [comment]
OK, I've dropped the rename kwargs commit. If that is still desirable I can move it to another ticket. This will change will simply cover the goals of the original ticket.

**Comment by timgraham** [comment]
merged in d861f95c445652e4269c9015836d14dcf8b9a587, thanks!

**Pull request closed without merging**